src/sidebar/components/tag-editor.js

```diff
@@ -1,35 +1,197 @@
 'use strict';
 
-// @ngInject
-function TagEditorController(tags) {
-  this.onTagsChanged = function() {
-    tags.store(this.tagList);
-
-    const newTags = this.tagList.map(function(item) {
-      return item.text;
-    });
-    this.onEditTags({ tags: newTags });
+const { createElement } = require('preact');
+const propTypes = require('prop-types');
+const { useRef, useState } = require('preact/hooks');
+
+const { withServices } = require('../util/service-context');
+const SvgIcon = require('./svg-icon');
+
+/**
+ * Component to edit annotation's tags.
+ */
+function TagEditor({ annotation, onEditTags, tags: tagsService, tagList }) {
+  const inputEl = useRef(null);
+  const [suggestions, setSuggestions] = useState([]);
+
+  /**
+   * Handle changes to this annotation's tags
+   */
+  const updateTags = tagList => {
+    // update suggested-tags list via service
+    tagsService.store(tagList.map(tag => ({ text: tag })));
+    // invoke callback with updated tags
+    onEditTags({ tags: tagList });
   };
 
-  this.autocomplete = function(query) {
-    return Promise.resolve(tags.filter(query));
+  /**
+   * Remove a tag from this annotation.
+   *
+   * @param {string} tag
+   */
+  const removeTag = tag => {
+    const newTagList = [...tagList]; // make a copy
+    const index = newTagList.indexOf(tag);
+    newTagList.splice(index, 1);
+    updateTags(newTagList);
   };
 
-  this.$onChanges = function(changes) {
-    if (changes.tags) {
-      this.tagList = changes.tags.currentValue.map(function(tag) {
-        return { text: tag };
+  /**
+   * Adds a tag equal to the value of the input field, and
```

Comment:
Good comment! Perhaps amend to `Add a tag to the annotation equal...` (i.e. specify the tag is getting added to an annotation).